### PR TITLE
Null query params

### PR DESCRIPTION
**Pull request by ArcherWheeler**
This is mostly just tests to make sure this works, which it does - sort of. The problem is for the query param types:

This is easy to implement (and works rn)
```
{ x: ?string }
```
This is hard to do (and not done)
```
{ x? : string }
```

Interestingly the server looking for this type will accept an input `url/foo?` without the param. However, the safe api client will throw an error. As discussed with @sleexyz irl, due to limitations of `$Exact<T>` and type splats there isn't an easy way to get the types to correctly merge _just_ for optionals.

#22

### CODE REVIEW

File: integration-tests/src/merge_query_params.test.js [hunk 1]
```diff
@@ -37,3 +37,58 @@ describe("for an endpoint with multiple queryString middleware", () => {
...
+
+    it("will remove undefined parameters", async () => {
+      const server = TestUtils.makeServer({
+        endpoint: testOptionalEndpoint,
+        handler: testOptionalHandler
+      });
+      const baseURL = `http://localhost:${server.address().port}`;
+      const input = { x: undefined };
+      await SafeFetch.safeGet(baseURL, testOptionalEndpoint, input);
+      const expectedURL = `${baseURL}/foo?`;
```
Comment (sleexyz) [inline integration-tests/src/merge_query_params.test.js]: should the expectedURL contain the `?` at all? Fine either way.

File: integration-tests/src/query_params.test.js [hunk 2]
```diff
@@ -93,6 +93,65 @@ describe("for a GET endpoint with no parameters", () => {
...
+> = SafeAPI.endpoint()
+  .fragment("foo")
+  .queryParams({
+    x: T.maybeString,
+    y: T.string
+  });
+
+const testOptionalHandler = async input => {
+  return {
+    x: input.x ? input.x.toString() : null,
```
Comment (sleexyz) [inline integration-tests/src/query_params.test.js]: `input.x || null`

File: integration-tests/src/query_params.test.js [hunk 3]
```diff
@@ -93,6 +93,65 @@ describe("for a GET endpoint with no parameters", () => {
...
+  });
+
+  test("server will accept absent optional inputs", async () => {
+    const resp = await fetch(
+      `http://localhost:${server.address().port}/foo?y=Y`
+    );
+    const json = await resp.json();
+
+    expect(json).toEqual({ x: null, y: "Y" });
+    expect(Object.keys(json)).toEqual(["x", "y"]);
```
Comment (sleexyz) [inline integration-tests/src/query_params.test.js]: redundant assertion

File: integration-tests/src/query_params.test.js [hunk 4]
```diff
@@ -93,6 +93,65 @@ describe("for a GET endpoint with no parameters", () => {
...
+    const resp = await fetch(
+      `http://localhost:${server.address().port}/foo?x=&y=Y`
+    );
+    const json = await resp.json();
+    expect(json).toEqual({ x: null, y: "Y" });
+  });
+
+  test("server will accept absent optional inputs", async () => {
+    const resp = await fetch(
+      `http://localhost:${server.address().port}/foo?y=Y`
```
Comment (sleexyz) [inline integration-tests/src/query_params.test.js]: Should we be testing with `safeFetch` instead? I would be curious what the deserialization of a serialized null looks like.

File: safe-api-fetch/src/index.js [hunk 5]
```diff
@@ -2,6 +2,7 @@
 import * as SafeAPI from "safe-api";
 import queryString from "querystring";
 import fetch from "isomorphic-fetch";
+import _ from "lodash";
```
Comment (sleexyz) [inline safe-api-fetch/src/index.js]: use vanilla for-loops (better typing, smaller bundle size, fewer ways to do things, fewer APIs to know to to read to codebase)

File: safe-api/src/type_rep.js [hunk 6]
```diff
@@ -2,4 +2,7 @@
 export interface TypeRep<T> {}
 
 class StringRep implements TypeRep<string> {}
+class MaybeStringRep implements TypeRep<?string> {}
+
 export const string = new StringRep();
+export const maybeString = new MaybeStringRep();
```
Comment (sleexyz) [inline safe-api/src/type_rep.js]: should `maybe` take a typerep and return a typerep, unioned with null and undefined? instead of encoding a `maybeFoo` for every type `Foo`.

File: integration-tests/src/merge_query_params.test.js (hunk 1 above)
Comment (ArcherWheeler) [reply integration-tests/src/merge_query_params.test.js]: so the query string builder outputs a `?` right now. I'm not against changing it, but this is a valid url.

File: integration-tests/src/query_params.test.js (hunk 4 above)
Comment (ArcherWheeler) [reply integration-tests/src/query_params.test.js]: I'd prefer to keep these unit tests, but I'll add a `foo?x=&y=Y` which is what we're testing for null serialization.

File: integration-tests/src/query_params.test.js (hunk 3 above)
Comment (ArcherWheeler) [reply integration-tests/src/query_params.test.js]: So... This is to explicitly test that the object contains `x` as a key with a null value. The expression `toEqual({ x: null, y: "Y" })` probably throws an error on `{y: "Y"}`, but I think this more clear of what's being tested from a doc point of view.

File: safe-api-fetch/src/index.js (hunk 5 above)
Comment (ArcherWheeler) [reply safe-api-fetch/src/index.js]: 🔨 :milk_glass: 🔥

File: integration-tests/src/merge_query_params.test.js (hunk 1 above)
Comment (astampoulis) [reply integration-tests/src/merge_query_params.test.js]: See my comment (I thought I posted it the other day but I see I didn't press submit). Moving the conditional should fix this. It's better to not output the `?` when there are no query params I think.

File: safe-api/src/type_rep.js (hunk 6 above)
Comment (ArcherWheeler) [reply safe-api/src/type_rep.js]: Yeah. Unless someone else is about to grab that I'd like that to be in the next PR. This is mostly to add a bunch of tests.

File: integration-tests/src/query_params.test.js (hunk 3 above)
Comment (astampoulis) [reply integration-tests/src/query_params.test.js]: Check to see whether `.toEqual({x: null, y: "Y"})` works with an object that does not contain `x` (I would guess it doesn't). If it doesn't I think the test above should be clear enough!

Comment (ArcherWheeler) [reply integration-tests/src/query_params.test.js]: never mind. This test is already written above.

File: integration-tests/src/query_params.test.js (hunk 4 above)
Comment (ArcherWheeler) [reply integration-tests/src/query_params.test.js]: never mind. This test is already written above.

Comment (astampoulis) [reply integration-tests/src/query_params.test.js]: :+1: Maybe move the test from `merge_query_params` here? That file is checking to see whether two calls to query params merge things correctly (esp. at the type level).

File: integration-tests/src/query_params.test.js [hunk 7]
```diff
@@ -93,6 +93,93 @@ describe("for a GET endpoint with no parameters", () => {
...
+      const json = await resp.json();
+      expect(json).toEqual({ x: "X", y: "Y" });
+    });
+
+    test("server will parse null inputs", async () => {
+      const resp = await fetch(
+        `http://localhost:${server.address().port}/foo?x=&y=Y`
+      );
+      const json = await resp.json();
+      expect(json).toEqual({ x: null, y: "Y" });
```
Comment (sleexyz) [inline integration-tests/src/query_params.test.js]: Right now, we're looking at response as returned by the server (as processed by the given handler). Instead, we should be testing what input the server handler sees as transparently as possible.

Do this by making the handler a spy:
```js
const handler = jest.fn(input => {
  ...
});
```
and then asserting `expect(handler).toHaveLastBeenCalledWith(...)`

Comment (sleexyz) [inline integration-tests/src/query_params.test.js]: Right now, we're looking at response as returned by the server (as processed by the given handler). Instead, we should be testing what input the server handler sees as transparently as possible.

Do this by making the handler a spy:
```js
const handler = jest.fn(input => {
  ...
});
```
and then asserting `expect(handler).toHaveLastBeenCalledWith(...)`

File: integration-tests/src/query_params.test.js [hunk 8]
```diff
@@ -93,6 +93,93 @@ describe("for a GET endpoint with no parameters", () => {
...
+    });
+
+    test("server will accept absent optional inputs", async () => {
+      const resp = await fetch(
+        `http://localhost:${server.address().port}/foo?y=Y`
+      );
+      const json = await resp.json();
+
+      expect(json).toEqual({ x: null, y: "Y" });
```
Comment (sleexyz) [inline integration-tests/src/query_params.test.js]: (See above comment, we should test input as the handler receives it as transparently as possible)

**Comment by ArcherWheeler** [comment]
@sleexyz It's a good thing you brought up the testing a spy rather than the response. I didn't realize "" is false-y (🙄 ) so `input.x || null` was overwriting ""s into nulls. This means that `nulls` are _not_ correctly preserved. I've added a test to show the behavior, which is the default node query string parsing behavior. I talked to @astampoulis irl, and we decided it'd be best to only override this behavior explicitly through typ_rep. That'll come in a later PR.
